crypto/crmf/crmf_lib.c

```diff
@@ -0,0 +1,767 @@
+             */
+            if (pubkey == NULL ||
+                sig->poposkInput->publicKey == NULL ||
+                X509_PUBKEY_cmp(pubkey, sig->poposkInput->publicKey) ||
+                ASN1_item_verify(ASN1_ITEM_rptr(OSSL_CRMF_POPOSIGNINGKEYINPUT),
+                                 sig->algorithmIdentifier, sig->signature,
+                                 sig->poposkInput,
+                                 X509_PUBKEY_get0(pubkey)) < 1)
+                break;
+        } else {
+            if (pubkey == NULL ||
+                req->certReq->certTemplate->subject == NULL ||
+                ASN1_item_verify(ASN1_ITEM_rptr(OSSL_CRMF_CERTREQUEST),
+                                 sig->algorithmIdentifier, sig->signature,
+                                 req->certReq,
+                                 X509_PUBKEY_get0(pubkey)) < 1)
+                break;
+        }
+        return 1;
+    case OSSL_CRMF_POPO_KEYENC:
+        if (req->popo->value.keyEncipherment->type
+            != OSSL_CRMF_POPOPRIVKEY_SUBSEQUENTMESSAGE)
+            goto unsupported;
+        if (ASN1_INTEGER_get
+            (req->popo->value.keyEncipherment->value.subsequentMessage)
+            != OSSL_CRMF_SUBSEQUENTMESSAGE_ENCRCERT)
+            goto unsupported;
+        /* TODO when implemented in CMP_certrep_new(): return 1 */
+        goto unsupported;
+    case OSSL_CRMF_POPO_KEYAGREE:
+    default:
+    unsupported:
+        CRMFerr(CRMF_F_OSSL_CRMF_MSGS_VERIFY_POPO,
+                CRMF_R_UNSUPPORTED_POPO_METHOD);
+        return 0;
+    }
+    CRMFerr(CRMF_F_OSSL_CRMF_MSGS_VERIFY_POPO,
+            CRMF_R_UNSUPPORTED_POPO_NOT_ACCEPTED);
+    return 0;
+}
+
+/* retrieves the serialNumber of the given cert template or NULL on error */
+ASN1_INTEGER *OSSL_CRMF_CERTTEMPLATE_get0_serialNumber(OSSL_CRMF_CERTTEMPLATE *tmpl)
+{
+    return tmpl != NULL ? tmpl->serialNumber : NULL;
+}
+
+/* retrieves the issuer name of the given cert template or NULL on error */
+X509_NAME *OSSL_CRMF_CERTTEMPLATE_get0_issuer(OSSL_CRMF_CERTTEMPLATE *tmpl)
+{
+    return tmpl != NULL ? tmpl->issuer : NULL;
+}
+
+/*-
+ * fill in certificate template.
+ * Any value argument that is NULL will leave the respective field unchanged.
+ */
+int OSSL_CRMF_CERTTEMPLATE_fill(OSSL_CRMF_CERTTEMPLATE *tmpl,
+                                const EVP_PKEY *pubkey,
+                                const X509_NAME *subject,
+                                const X509_NAME *issuer,
+                                const ASN1_INTEGER *serial)
+{
+    if (tmpl == NULL) {
+        CRMFerr(CRMF_F_OSSL_CRMF_CERTTEMPLATE_FILL, CRMF_R_NULL_ARGUMENT);
+        return 0;
+    }
+    if (pubkey != NULL && !X509_PUBKEY_set(&tmpl->publicKey, (EVP_PKEY *)pubkey))
+        goto oom;
+    if (subject != NULL && !X509_NAME_set(&tmpl->subject, (X509_NAME *)subject))
+        goto oom;
+    if (issuer != NULL && !X509_NAME_set(&tmpl->issuer, (X509_NAME *)issuer))
```

Comment:
Are all the casts here safe? It's not clear to me if they are or not. I started to look but rapidly started going down a rabbit hole. Either we need to fix the underlying functions to declare themselves as taking const params, and explore the rabbit hole properly - or we have to assume that it isn't safe.

Reply:
The idea was that all structures pointed to by parameters (except the first one) of `OSSL_CRMF_CERTTEMPLATE_fill()` should not be touched and this should be reflected by declaring them `const *`. Yet it turns out that `X509_PUBKEY_set()` for whatever reason increments the reference counter of the public key rather than copying the structure. The other parameters can be safely considered constant because `X509_NAME_set()` and `ASN1_INTEGER_dup()` actually do not touch them. Since I found it rather unfortunate that these functions are not (yet) constified I started doing that.

In the end I've constified all *`_dup()` functions as far as possible, as well as most of *`i2d_`*`()` and related functions. This turned out to be quite some amount of work, hopefully to the benefit of many. Among others, I had to introduce a couple of small helper functions and declarations deep inside the ASN.1 implementations.

Reply:
I fully support the API constification you made, but it deserves its own PR, IMHO, as it is large patch and not dependent of CRMF support added in this PR.

Reply:
I agree. How best to deal with the problem that this (CRMF) PR will depend on that API constification PR?

Reply:
Create a separate topic branch for the constification, cherry-pick the constification commit, and create a pull request for it.

Then rebase this branch on top of the constification topic branch. Git should notice automatically that the constification commit was already applied. Alternatively, do an interactive rebase (`git rebase -i`) and remove the constification commit from the rebase-todo-list.

You should add a note to the description of this pr that it is based on the other pr, and optionally add the `WIP: ` prefix to the pr title while it's not based onto master.

After the other constification pull request is accepted and merged, you can rebase back onto master.

If you need more technical assistance with the details, just drop me a note.

HTH,
Matthias

Reply:
Thanks @mspncp for this hit!
I'll follow it as soon as this PR has been rebased on the current master.